core/src/main/scala/org/apache/spark/CacheManager.scala

```diff
@@ -136,7 +147,13 @@ private[spark] class CacheManager(blockManager: BlockManager) extends Logging {
        */
       blockManager.putIterator(key, values, level, tellMaster = true, effectiveStorageLevel)
       blockManager.get(key) match {
-        case Some(v) => v.data.asInstanceOf[Iterator[T]]
+        case Some(v) =>
```

Comment:
Yep, that's right. Any lock acquired in a task and not explicitly released by that task will be auto-released upon task completion or failure.